sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/BuiltinApproxCountDistinctSqlAggregator.java

```diff
@@ -118,12 +120,14 @@ public Aggregation toDruidAggregation(
       }
 
       if (inputType.is(ValueType.COMPLEX)) {
-        aggregatorFactory = new HyperUniquesAggregatorFactory(
-            aggregatorName,
-            dimensionSpec.getOutputName(),
-            false,
-            true
-        );
+        if (validateInputType(inputType)) {
```

Comment:
why not do the `return null` here with the setPlanning error?

Reply:
looking at the variables: `inputAccessor`, `arg`, `rexNode`, `dataType` - they are "connected":
* the 1st if uses a natively converted expressions directcolumn property to look it back up in the inputaccessor
* the else branch was going the other way and was asking for the type of the input argument
* if the 1st argument was converted to an expression which directly read an input column; then that means the `rexNode`-s type must have been the same as `dataType` unless it couldn't have been a direct column access

![gif](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExY3YyN2VlaTIyYWhtbG9vcXM0Z2Q2bmJhNnpoaWpzZ2tuaGV1MzZ3byZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/39xDerRCIoX2WeUVBz/giphy.gif)

so...
```
inputAccessor.getInputRowSignature()
            .getColumnType(arg.getDirectColumn())
```
is essentially the same but is only accurate if its a `DirectColumn` ; so 
you could move this check outside alongside with:
```
      final RelDataType dataType = rexNode.getType();
final ColumnType inputType = Calcites.getColumnTypeForRelDataType(dataType);
```
and make it only there...

Reply:
Made the changes as discussed offline. Thanks for the help and guidance!